File: app/src/main/java/com/mxt/anitrend/adapter/recycler/index/EpisodeAdapter.java
package com.mxt.anitrend.adapter.recycler.index;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;

import com.annimon.stream.Stream;
import com.bumptech.glide.Glide;
import com.mxt.anitrend.R;
import com.mxt.anitrend.base.custom.recycler.RecyclerViewAdapter;
import com.mxt.anitrend.base.custom.recycler.RecyclerViewHolder;
import com.mxt.anitrend.databinding.AdapterEpisodeBinding;
import com.mxt.anitrend.model.entity.crunchy.Episode;

import java.util.ArrayList;
import java.util.List;

import butterknife.OnClick;
import butterknife.OnLongClick;

/**
 * Created by max on 2017/11/04.
 */

public class EpisodeAdapter extends RecyclerViewAdapter<Episode> {

    public EpisodeAdapter(Context context) {
        super(context);
    }

    @NonNull
    @Override
    public RecyclerViewHolder<Episode> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new EpisodeViewHolder(AdapterEpisodeBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
    }

    @Override
    public Filter getFilter() {
        return null;
    }

    protected class EpisodeViewHolder extends RecyclerViewHolder<Episode> {

        private AdapterEpisodeBinding binding;

        /**
         * Default constructor which includes binding with butter knife
         *
         * @param binding
         */
        public EpisodeViewHolder(AdapterEpisodeBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        /**
         * Load image, text, buttons, etc. in this method from the given parameter
         * <br/>
         *
         * @param model Is the model at the current adapter position
         */
        @Override
        public void onBindViewHolder(Episode model) {
            binding.setModel(model);
            binding.setPresenter(presenter);
            binding.executePendingBindings();
        }

        /**
         * If any image views are used within the view holder, clear any pending async img requests
         * by using Glide.clear(ImageView) or Glide.with(context).clear(view) if using Glide v4.0
         * <br/>
         *
         * @see Glide
         */
        @Override
        public void onViewRecycled() {
            Glide.with(getContext()).clear(binding.seriesImage);
            binding.unbind();
        }

        @OnClick(R.id.series_image)
        @Override
        public void onClick(View v) {
            performClick(clickListener, data, v);
        }

        @Override @OnLongClick(R.id.series_image)
        public boolean onLongClick(View v) {
            return performLongClick(clickListener, data, v);
        }
    }
}

File: app/src/main/java/com/mxt/anitrend/adapter/recycler/index/EpisodeAdapter.kt
package com.mxt.anitrend.adapter.recycler.index

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import com.bumptech.glide.Glide
import com.mxt.anitrend.R
import com.mxt.anitrend.base.custom.recycler.RecyclerViewAdapter
import com.mxt.anitrend.base.custom.recycler.RecyclerViewHolder
import com.mxt.anitrend.base.custom.view.image.WideImageView
import com.mxt.anitrend.databinding.AdapterEpisodeBinding
import com.mxt.anitrend.extension.getLayoutInflater
import com.mxt.anitrend.model.entity.crunchy.Episode
import java.util.Locale
import java.util.concurrent.TimeUnit

/**
 * Created by max on 2017/11/04.
 */
class EpisodeAdapter(
    context: Context,
) : RecyclerViewAdapter<Episode>(context) {

    private fun formatDuration(duration: String?): String {
        val timeSpan = duration?.toLongOrNull() ?: return "00:00"
        val minutes = TimeUnit.SECONDS.toMinutes(timeSpan)
        val seconds = timeSpan - TimeUnit.MINUTES.toSeconds(minutes)
        return String.format(Locale.getDefault(), if (seconds < 10) "%d:0%d" else "%d:%d", minutes, seconds)
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): RecyclerViewHolder<Episode> = EpisodeViewHolder(
        AdapterEpisodeBinding.inflate(parent.context.getLayoutInflater(), parent, false),
    )

    override fun getFilter(): Filter? = null

    inner class EpisodeViewHolder(
        private val binding: AdapterEpisodeBinding,
    ) : RecyclerViewHolder<Episode>(binding.root) {
        init {
            bindClickListeners(R.id.series_image)
            bindLongClickListeners(R.id.series_image)
        }

        override fun onBindViewHolder(model: Episode) {
            WideImageView.setImage(binding.seriesImage, model.thumbnail?.getOrNull(0)?.url)
            binding.seriesDuration.text = formatDuration(model.content?.duration)
            binding.seriesTitle.text = model.title
        }

        override fun onViewRecycled() {
            Glide.with(getContext()).clear(binding.seriesImage)
        }

        override fun onClick(v: View) {
            performClick(clickListener, data, v)
        }

        override fun onLongClick(v: View): Boolean = performLongClick(clickListener, data, v)
    }
}
